fold 'kola testiso' into 'kola run'

nikita-dubrovskii opened the pull request:
This huge PR folds all `testiso` tests:
```
$ kola run iso.* --parallel 4 --inst-insecure                                
=== RUN   iso.as-disk.uefi                          
=== RUN   iso.miniso-install.nm                   
=== RUN   iso.miniso-install.4k.uefi                                                                                                                                                                               
=== RUN   iso.pxe-offline-install.rootfs-appended
=== RUN   iso.live-login.uefi                                                                            
=== RUN   iso.offline-install-iscsi.ibft.uefi                                                                                                                                                                      
=== RUN   iso.as-disk.uefi-secure                                                                        
=== RUN   iso.fips.uefi                                                                                  
=== RUN   iso.pxe-online-install               
=== RUN   iso.live-login                                                                                 
=== RUN   iso.miniso-install.4k.nm.uefi           
=== RUN   iso.install                        
=== RUN   iso.offline-install-fromram.4k.uefi
=== RUN   iso.offline-install-iscsi.ibft-with-mpath
=== RUN   iso.offline-install-iscsi.manual
=== RUN   iso.miniso-install
=== RUN   iso.offline-install
=== RUN   iso.pxe-online-install.4k.uefi
=== RUN   iso.live-login.uefi-secure
=== RUN   iso.as-disk
=== RUN   iso.pxe-offline-install.4k.uefi
=== RUN   iso.offline-install.mpath
--- PASS: iso.as-disk.uefi (37.11s)
--- PASS: iso.live-login.uefi-secure (39.32s)
--- PASS: iso.as-disk.uefi-secure (37.72s)
--- PASS: iso.offline-install-fromram.4k.uefi (92.64s)
--- PASS: iso.install (89.74s)
--- PASS: iso.offline-install.mpath (129.49s)
--- PASS: iso.live-login (55.92s)
--- PASS: iso.fips.uefi (49.65s)
--- PASS: iso.miniso-install.4k.nm.uefi (112.01s)
--- PASS: iso.as-disk (36.91s)
--- PASS: iso.pxe-online-install (92.90s)
--- PASS: iso.pxe-offline-install.4k.uefi (97.06s)
--- PASS: iso.miniso-install (94.49s)
Warning: Cannot announce submounts, client does not support it
--- PASS: iso.pxe-online-install.4k.uefi (82.69s)
--- PASS: iso.offline-install (100.28s)
--- PASS: iso.live-login.uefi (24.93s)
Warning: Cannot announce submounts, client does not support it
--- PASS: iso.pxe-offline-install.rootfs-appended (120.32s)
--- PASS: iso.miniso-install.4k.uefi (82.57s)
--- PASS: iso.offline-install-iscsi.ibft.uefi (174.86s)
--- PASS: iso.miniso-install.nm (94.50s)
--- PASS: iso.offline-install-iscsi.ibft-with-mpath (166.62s)
--- PASS: iso.offline-install-iscsi.manual (151.04s)
PASS, output in tmp/kola/qemu-2025-11-24-1353-1209

```

https://github.com/coreos/coreos-assembler/issues/3989

nikita-dubrovskii committed:
mantle/platform/qemu: add MachineBuilder for customizable machine creation

Introduce MachineBuilder struct to allow callers to customize QEMU
machine creation by providing hooks for builder initialization, disk
setup, network configuration.

This will be used to fold 'testiso' tests into the standard 'kola run',
where custom machine creation logic is needed for ISO-based
installations.

nikita-dubrovskii committed:
kola/tests/iso: fold testiso live-login.* tests into 'kola run'

This migrates the iso-live-login tests from the standalone 'kola testiso'
command into the standard 'kola run' framework, making them available
alongside other kola tests.

Changes:
- Moved iso-live-login* tests implementation from cmd/kola/testiso.go to
  mantle/kola/tests/iso/live-login.go
- Removed 4k test variants (iso-live-login.4k.uefi) as they were
  non-functional for this test type

Why 4k tests were dropped:
The 4k flag in testiso only affects disk sector size during disk creation.
However, the live-login test boots directly from ISO without creating any
disks, so the 4k flag had no effect. The 4k test variants were redundant
and ran identical code to their non-4k counterparts.

nikita-dubrovskii committed:
mantle/platform: move metal.go to machine/qemu/

Relocate metal.go to machine/qemu/ directory where it belongs, as it
contains QEMU-specific code for ISO installations and is only used by
ISO tests, not as a general platform implementation.

nikita-dubrovskii committed:
mantle/platform/machine/qemu: merge InstalledMachine into machine struct

Temporarily consolidate InstalledMachine into machine struct to simplify
the codebase during the testiso migration. This prepares for migrating ISO
installation tests from the testiso command into the standard 'kola run'
test framework.

This merge is temporary and will be refactored once the migration is complete.

nikita-dubrovskii committed:
kola/tests/iso: fold testiso (iso|miniso)-(offline-)?install* tests

nikita-dubrovskii committed:
kola/tests/iso: fold testiso *pxe* tests

nikita-dubrovskii committed:
kola/tests/iso: fold testiso *iscsi* tests

nikita-dubrovskii committed:
kola/tests/iso: fold testiso iso-as-disk tests

nikita-dubrovskii committed:
kola/tests/iso: refactor live-iso.go into independent test files

Split the monolithic live-iso.go file into several focused files where
each file contains a specific set of tests:

- live-pxe.go: PXE network boot tests
- live-iscsi.go: iSCSI boot and installation tests
- live-as-disk.go: ISO-as-disk installation tests
- common.go: Shared test utilities and helper functions

This improves code organization and makes it easier to locate and
maintain specific test suites.

nikita-dubrovskii committed:
mantle/cmd/kola: move inst-insecure and pxe-kargs to global QEMUOptions

Move the 'inst-insecure' and 'pxe-kargs' parameters from local test options
to global QEMUOptions structure, making them available as command-line
flags for all ISO tests:

- --inst-insecure: Skip signature verification on metal image
- --pxe-kargs: Additional kernel arguments for PXE boot tests

nikita-dubrovskii committed:
kola/tests/iso: fold testiso iso-fips.uefi test

Migrate the last remaining test from the legacy 'kola testiso'
command into the standard 'kola run' This completes the migration
of all ISO tests and removes the now-obsolete testiso.go file.

nikita-dubrovskii committed:
kola/tests/iso: refactor *fips* tests

Update the test to use the existing MachineBuilder API after the
testiso migration.

nikita-dubrovskii committed:
kola/tests/iso: refactor *iscsi* tests

nikita-dubrovskii committed:
kola/tests/iso: extract CheckTestOutput helper function

Extract duplicate completion channel reading logic into a reusable
CheckTestOutput helper function. This reduces code duplication across
iso tests.

nikita-dubrovskii committed:
kola/tests/iso: remove redundant console and journal checks

The kola test harness already performs console and journal checks
automatically after each test completes, making the duplicate checks
in awaitCompletion unnecessary.

Call flow in mantle/kola/harness.go:
```
  runProvidedTests
    -> runTest
         -> handleConsoleChecks("console", ...)
              -> CheckConsole
         -> handleConsoleChecks("journal", ...)
              -> CheckConsole
```

nikita-dubrovskii committed:
kola/tests/iso: refactor iso-as-disk* tests

Also remove the 4k sector size test variant as it's technically impossible
to configure. The AddIso() function and iso-as-disk implementation do
not support sector size configuration - the ISO is attached as a raw,
readonly device without any sector size options.

The Disk.SectorSize field only applies to regular disk images added
via AddDisk() or AddDisksFromSpecs(), not to ISOs attached via AddIso().
Since iso-as-disk tests boot directly from the ISO file (not a disk
image), sector size configuration is not applicable.

nikita-dubrovskii committed:
kola/tests/iso: refactor (iso|miniso)-(offline-)?install* tests

nikita-dubrovskii committed:
kola/tests/iso: refactor *pxe* tests

nikita-dubrovskii committed:
mantle/platform/machine/qemu: drop no longer used code

Remove deprecated machine struct fields and methods that are no longer
needed after the migration all ISO tests under kola.

nikita-dubrovskii commented:
Finished metal PXE&ISO installation. Almost ready for review

dustymabe commented:
@nikita-dubrovskii this is an incredible amount of work. Thank you for taking this on.

This does make it hard to grok all in one go, though. I'm still struggling a bit with the first commit even. As I go I'm questioning old code that we have and I want to challenge if some of it is necessary.

I broke out some initial commits that challenge the original code (and also a few commits from this PR) into https://github.com/coreos/coreos-assembler/pull/4486

I also rebased the commits from this PR on top of that in https://github.com/dustymabe/coreos-assembler/commits/dusty-kola-testiso-fold/ if you're interested in a rebased version of this PR.

Let me know what you think.

nikita-dubrovskii commented:
@dustymabe Thx, checked your PR. Looks sane to me, and many thanks for better commit messages, describing purpose of changes. The only thing holding me back is dealing with rebase&merge conflicts, but not a big deal.

nikita-dubrovskii committed:
mantle/qemu: conditionally wait for SSH address

Only wait for SSH address when usermode networking is enabled.
This allows ISO tests without networking to proceed without
unnecessary SSH address waiting.

nikita-dubrovskii committed:
mantle/qemu: add Instance() method to access QemuInstance

Add Instance() method to expose the underlying QemuInstance from
a platform.Machine. This allows tests to access QEMU-specific
functionality not available through the generic interface.

nikita-dubrovskii committed:
kola/tests/iso: Dump journal on emergency.target in ISO tests

Add journal dumping to signalFailureUnit to capture diagnostic
information when the system enters emergency.target during ISO tests.

After switch root occurs during live ISO testing, coreos-installer runs
from the real root filesystem rather than the initramfs, which means
ignition-virtio-dump-journal.service is no longer enabled. This change
ensures that when emergency.target is reached, the journal is still
dumped to the virtio port, allowing platform.StartMachine to capture errors.

The unit now checks if ignition-virtio-dump-journal.service is enabled,
and if not, manually executes the journal dump script from the dracut
emergency shell setup module.

nikita-dubrovskii committed:
mantle/platform: Add IgnitionPath() helper to QemuBuilder

Extract the Ignition config path logic into a reusable IgnitionPath()
method that returns the path where the Ignition config will be written.

This allows callers to get the Ignition config path before rendering,
which is useful for tests that need to create symlinks or perform other
operations on the config file path before the machine is started.

nikita-dubrovskii committed:
mantle/platform: Allow passing nil instead of Ignition config

Allow callers to pass nil userdata/config when Ignition configuration
is not needed, such as in PXE boot scenarios where configuration is
provided through alternative means.

nikita-dubrovskii committed:
mantle/kola/tests/iso: Use firmware string instead of booleans

Replace enableUefi and enableUefiSecure boolean flags with a single
firmware string field in IsoTestOpts for simpler and more consistent
firmware configuration across all ISO tests.

nikita-dubrovskii committed:
mantle/kola/tests/iso: Remove obsolete console field from IsoTestOpts

dustymabe reviewed (commented):
I spent a good amount of time (most of my day really) reviewing this PR. I'm not going to lie it's really hard to review. I first started trying each individual commit. Then I tried just looking at the final result and diffing with the original testiso code, which wasn't really much easier because so much stuff moved around.

At the end of the day here we've managed to share some more code but I really don't know if we've improved the ability to read/comprehend things and it kind of makes me sick to my stomach. One example is I'm not sure how likely we are to miss something in the code review here and introduce a regression to the testing.

Since you wrote the code in this PR (and have a better understanding) do you find it easier to read now? Or do you also think it's still hard to follow? Don't get me wrong the original testiso code isn't easy to follow either so that wouldn't be a reason for not merging this.

The two files that are hardest for me right now are live-pxe.go and live-iso.go, but that's probably mostly because of the sheer amount of code in them.

nikita-dubrovskii committed:
mantle/platform/qemu: add setter for netdev's bootindex property

Add SetNetbootIndex() method to QemuBuilder to allow configuring the
bootindex property for network devices. This enables proper boot order
control when using PXE/network boot, as an alternative to the legacy
-boot order=n option.

The bootindex property is mutually exclusive with -boot order, as most
firmware implementations support one or the other but not both. When
bootindex is set, the -boot order option is automatically omitted.

nikita-dubrovskii committed:
kola/tests/iso: simplify iso.pxe* networking initialization

Simplify PXE test network setup by using QemuBuilder's built-in methods
(SetNetbootP, SetNetbootIndex, EnableUsermodeNetworking) instead of
manually constructing QEMU network device arguments.

nikita-dubrovskii commented on mantle/kola/tests/iso/live-pxe.go:
```diff
@@ -0,0 +1,512 @@
+		}
+		builder.Append("-netdev", usernetdev)
+		// for SSH access
+		return qc.SetupDefaultNetwork(o, builder)
```
Nope, before systemd untits forced VM to `poweroff` and code checked only virtio channels.  Having ssh here makes tests more `kola` style. Btw, i've further reworked this code

nikita-dubrovskii committed:
kola/tests/iso: add EnsureLiveArtifactsExist helper function for ISO tests

Add a helper function that skips tests if no artifacts are present.

nikita-dubrovskii commented on mantle/kola/tests/iso/live-login.go:
```diff
@@ -0,0 +1,118 @@
+
+var (
+	tests_live_login_x86_64 = []string{
+		"live-login",
```
done

nikita-dubrovskii commented on mantle/kola/tests/iso/live-login.go:
```diff
@@ -0,0 +1,118 @@
+	}
+	tests_live_login_aarch64 = []string{
+		"live-login",
+		"live-login.uefi",
```
done

nikita-dubrovskii commented on mantle/kola/tests/iso/live-iscsi.go:
```diff
@@ -0,0 +1,197 @@
+	config.CopyKeys(keys)
+	// Add a failure target to stop the test if something go wrong rather than waiting for the 10min timeout
+	config.AddSystemdUnit("coreos-test-entered-emergency-target.service", signalFailureUnit, conf.Enable)
+	config.MountHost("/var/cosaroot", true)
```
Makes no sense anymore

nikita-dubrovskii commented on mantle/kola/tests/iso/common.go:
```diff
@@ -359,7 +364,11 @@ DefaultDependencies=false
 Type=oneshot
 RemainAfterExit=yes
-ExecStart=/bin/sh -c '/usr/bin/echo %s >/dev/virtio-ports/testisocompletion && systemctl poweroff'
+ExecStart=/bin/sh -c '/usr/bin/echo %s >/dev/virtio-ports/testisocompletion'
+ExecStart=/bin/bash -c '\
+if ! systemctl is-enabled ignition-virtio-dump-journal.service >/dev/null 2>&1; then \
+  exec /usr/lib/dracut/modules.d/99emergency-shell-setup/ignition-virtio-dump-journal.sh; \
+fi'
```
Summary from private discussions: indeed thru channels we know whether or not tests succeeded, but after `SkipStartMachine` was dropped, here is risk to hang - `platform.StartMachine()` can be blocked when `coreos-installer` failed (in real root), because there is no `IgnitionError()` (initramfs) and also it's not possible to `ssh`. So dumping failures to `ignition-virtio` chanell is a workaround here.

nikita-dubrovskii committed:
kola/tests/iso: align test function names to testLive* pattern

nikita-dubrovskii commented:
/gemini review

dustymabe commented on src/cmd-kola:
```diff
@@ -19,6 +19,49 @@ from cosalib import cmdlib
 basearch = cmdlib.get_basearch()
 
+
+def translate_testiso_args(kolaargs, subargs, unknown_args) -> bool:
```
I think maybe the better approach here is to not try to keep backwards compatibility with the old setup.

Instead WDYT about:

1. we adjust pipeline to build ISO artifacts up front (same time as qemu).. actually we probably just go ahead and build all artifacts up front?
2. we adjust pipeline code to check if `kola testiso` exists before calling it.
3. we update the code here in this PR to [`SKIP`](https://github.com/coreos/coreos-assembler/blob/211dee611ec96ca39bff31f70bac80c6f5e81d26/mantle/kola/tests/ignition/luks.go#L198) iso tests if no iso/live artifacts exist. (not sure how hard this would be).

The outcomes:

1. will allow us to run iso tests along with our other tests via `kola run` (i.e. this represents the modernization).
2. will allow older branches to still run the testiso code (because we didn't backport this PR to older RHCOS releases/COSA)
3. will allow people who are doing local development and just built the qemu to not have a fatal outcome.

nikita-dubrovskii replied:
Ok. Will do

dustymabe commented on mantle/kola/tests/iso/common.go:
```diff
@@ -0,0 +1,407 @@
+ExecStart=/bin/bash -c '\
+if ! systemctl is-enabled ignition-virtio-dump-journal.service >/dev/null 2>&1; then \
+  exec /usr/lib/dracut/modules.d/99emergency-shell-setup/ignition-virtio-dump-journal.sh; \
+fi'
```
I still have my reservations about this, but have resolved to leave fixing it to a followup.

One comment on the current approach: Since this unit doesn't run in the initramfs there's no way (IIUC) that `ignition-virtio-dump-journal.service` will ever be enabled (it only runs in the initramfs), so the `if` should never be false. Should be able to drop the if completely and just exec the shell script unconditionally.

nikita-dubrovskii replied:
Done.

nikita-dubrovskii committed:
kola/tests/iso: fix dev build detection pattern

The dev build naming pattern now uses '.devN' or '-devN' instead of '.dev.N' format:
  fedora-coreos-43.20260429.20.dev1-qemu.x86_64.qcow2
  rhcos-10.2.20260428-dev0-qemu.x86_64.qcow2

nikita-dubrovskii committed:
kola/tests/iso: Skip base checks until bootupd fix lands

Add SkipBaseChecksTag to all ISO live tests to temporarily skip base
checks that look at the journal for failures. This is needed until the
bootupd fix is merged.

Related: https://github.com/coreos/fedora-coreos-tracker/issues/2136

dustymabe commented on mantle/kola/tests/iso/iscsi_butane_setup.yaml:
```diff
@@ -180,3 +180,7 @@ systemd:
+    # Mask until bootupd fix lands
+    # https://github.com/coreos/fedora-coreos-tracker/issues/2136
+    - name: bootloader-update.service
+      mask: true
```
This shouldn't be needed.

I ran a local test with it deleted and the test did not fail.

nikita-dubrovskii replied:
Hmm, on my local build this was required. I'll recheck

nikita-dubrovskii replied:
Indeed,  dropped masking

dustymabe commented:
This LGTM if we get the masking of bootloader-update.service in iscsi_butane_setup.yaml removed.

nikita-dubrovskii committed:
cci: drop legacy testiso and build artifacts before kola

nikita-dubrovskii committed:
kola/tests/iso: add reprovision tag to live tests

Mark the live ISO tests with the reprovision tag because those
require a lot of memory and should run serially.
